### DISCUSSION TITLE
Retry-After header missing?

### DISCUSSION THREAD

**Discussion by ewels** [community/community#24760] [Apps, API and Webhooks]
<p>If you hit the GitHub API with a lot of requests at once to make PRs etc, you may run in to the API <a href="https://docs.github.com/en/rest/overview/resources-in-the-rest-api#abuse-rate-limits" rel="noopener nofollow ugc">Abuse rate limits</a>.</p>
<p>The documentation tells us that:</p>
<blockquote>
<p>When you have been limited, use the <code>Retry-After</code> response header to slow down. The value of the <code>Retry-After</code> header will always be an integer, representing the number of seconds you should wait before making requests again. For example, <code>Retry-After: 30</code> means you should wait 30 seconds before sending more requests.</p>
</blockquote>
<p>However, when I run in to these problems I never seem to have that header set:</p>
<pre><code class="lang-auto">{
	"message": "You have triggered an abuse detection mechanism and have been temporarily blocked from content creation. Please retry your request again later.",
	"documentation_url": "https://docs.github.com/overview/resources-in-the-rest-api#abuse-rate-limits"
}
</code></pre>
<pre><code class="lang-auto">{
	'Server': 'GitHub.com',
	'Date': 'Tue, 11 May 2021 11:10:25 GMT',
	'Content-Type': 'application/json; charset=utf-8',
	'Transfer-Encoding': 'chunked',
	'X-OAuth-Scopes': 'repo, workflow',
	'X-Accepted-OAuth-Scopes': '',
	'X-GitHub-Media-Type': 'github.v3; format=json',
	'X-RateLimit-Limit': '5000',
	'X-RateLimit-Remaining': '4628',
	'X-RateLimit-Reset': '1620734914',
	'X-RateLimit-Used': '372',
	'X-RateLimit-Resource': 'core',
	'Access-Control-Expose-Headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset',
	'Access-Control-Allow-Origin': '*',
	'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload',
	'X-Frame-Options': 'deny',
	'X-Content-Type-Options': 'nosniff',
	'X-XSS-Protection': '0',
	'Referrer-Policy': 'origin-when-cross-origin, strict-origin-when-cross-origin',
	'Content-Security-Policy': "default-src 'none'",
	'Vary': 'Accept-Encoding, Accept, X-Requested-With',
	'Content-Encoding': 'gzip',
	'X-GitHub-Request-Id': '0780:11AC:3AA0BD0:6B7151F:609A6620'
}
</code></pre>
<p>The <code>Access-Control-Expose-Headers</code> lists <code>Retry-After</code> as an allowed header, but the <code>Retry-After</code> header itself is not there.</p>
<p>I’m calling the API using the <code>requests</code> library in Python, code <a href="https://github.com/nf-core/tools/blob/7721f3742fc30a9e6e46d879404bd1366345092d/nf_core/sync.py#L348-L352" rel="noopener nofollow ugc">here</a>.</p>
<p>Am I missing something at my end, or is this a problem with the GitHub API?</p>
<p>Phil</p>

**Comment by nethgato** [comment, marked as answer]
<p>Hi there <a class="mention" href="/ewels">@ewels</a> :wave:</p>
<p>Thanks so much for joining us and for sharing your experience. There is additional documentation that speaks to the lack of <code>Retry-After</code> header found here:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-abuse-rate-limits" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-abuse-rate-limits" target="_blank" rel="noopener">Best practices for integrators - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-abuse-rate-limits" target="_blank" rel="noopener">//docs.github.com/en/rest/guides/best-practices-for-integrators</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Specifically:</p>
<blockquote>
<ul>
<li>Requests that create content which triggers notifications, such as issues, comments and pull requests, may be further limited and will not include a <code>Retry-After</code> header in the response. Please create this content at a reasonable pace to avoid further limiting.</li>
</ul>
</blockquote>
<p>The behavior you described in your OP:</p>
<blockquote>
<p>If you hit the GitHub API with a lot of requests at once to make PRs etc</p>
</blockquote>
<p>…is precisely what we try to protect against. As this is an abuse protection measure, it is based on multiple input variables, and there isn’t an X value of seconds to suggest.</p>
<p>The process to resolve this would be to decrease the frequency of the scripts calls dramatically, and increase as needed. It looks like the code already has logic to identify an updated frequency value, when that header is present; awesome!</p>
<p>Just gotta slow down to start, and everything should behave as you’re hoping.</p>

**Reply by tyeth** [reply]
Hey, I'm now regularly not getting the rate limit info included in responses. Presumably more anti-hammering, I don't quite get it as I'm still able to make more requests? All the rate related headers are missing except token expiration:
```
-- Remaining defaulted to -1: Github responded with no rate limit info, possible problems, printing reponse (200):
Request (get) - URL: /repos/adafruit/Adafruit_DRV8830/releases
Response text: []
Response header Date=Tue, 01 Oct 2024 18:50:30 GMT
Response header Content-Type=application/json; charset=utf-8
Response header Content-Length=2
Response header Cache-Control=private, max-age=60, s-maxage=60
Response header Vary=Accept, Authorization, Cookie, X-GitHub-OTP,Accept-Encoding, Accept, X-Requested-With
Response header ETag="cbff93ad9ca68eafbd30c57f01c1049f36162ed46ab7a757a38674943a9e85e0"
Response header X-OAuth-Scopes=read:org, read:packages, read:public_key, read:repo_hook, read:user, repo, user:email, workflow
Response header X-Accepted-OAuth-Scopes=repo
Response header github-authentication-token-expiration=2024-10-01 23:00:00 UTC
Response header X-GitHub-Media-Type=github.v3; param=hellcat-preview; format=json
Response header x-github-api-version-selected=2022-11-28
Response header Access-Control-Expose-Headers=ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset
Response header Access-Control-Allow-Origin=*
Response header Strict-Transport-Security=max-age=31536000; includeSubdomains; preload
Response header X-Frame-Options=deny
Response header X-Content-Type-Options=nosniff
Response header X-XSS-Protection=0
Response header Referrer-Policy=origin-when-cross-origin, strict-origin-when-cross-origin
Response header Content-Security-Policy=default-src 'none'
Response header Server=github.com
Response header X-GitHub-Request-Id=ADF4:1C3425:9115B8:99D8E8:66FC4476
-- Continuing as if nothings wrong 😇
```

![image](https://github.com/user-attachments/assets/3af1962c-5856-4d68-ac52-bda3c3133b4d)

**Comment by ewels** [comment]
<p>Amazing, thank you for the comprehensive answer <a class="mention" href="/nethgato">@nethgato</a>! I’m not quite sure how I missed that, seeing as the bullet is the one below the one I pasted in my original post :man_facepalming: I think I must have missed the critical word <em>not</em> (include…).</p>
<p>Anyway, that makes sense and as you say I added a backup approach if the header isn’t there. The context is a sync that runs once every few months but does various repos in parallel so submits 40 odd PRs all at once. How synchronous that is depends on how fast GitHub Actions runs the jobs so the abuse mechanism varies a lot. In the last run it was just  one PR, in the one before it was about half. Anyway, my random delay seemed to work on the most recent run so I’m happy :+1:</p>
<p>The main thing was that I thought I was going crazy / doing something wrong! So thank you for putting my mind at rest :slight_smile:</p>

**Comment by nethgato** [comment]
<p>So glad to help! You’re definitely not going crazy and I’m happy I was able to help verify sanity.</p>
<p>:bow:</p>

**Reply by Deepakgthomas** [reply]
I have a slightly similar [question](https://github.com/orgs/community/discussions/43385). I would highly appreciate your feedback.
